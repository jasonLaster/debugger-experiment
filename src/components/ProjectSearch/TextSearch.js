// @flow

import React, { Component, DOM as dom, PropTypes } from "react";
import classnames from "classnames";

import escapeRegExp from "lodash/escapeRegExp";
import Svg from "../shared/Svg";

import ManagedTree from "../shared/ManagedTree";
import SearchInput from "../shared/SearchInput";

import "./TextSearch.css";

import { getRelativePath } from "../../utils/sources-tree";

import type { Match } from "../../utils/search/project-search";
import type { Item } from "../shared/ManagedTree";

type Result = {
  filepath: string,
  matches: Match[]
};

type Results = Result[];

type Props = {
  results: Results,
  query: string,
  closeActiveSearch: () => void,
  searchSources: () => void,
  selectSource: () => void,
  searchBottomBar: any
};

type State = {
  inputValue: string,
  focused: boolean
};

export default class TextSearch extends Component {
  props: Props;
  state: State;

  constructor(props: Props) {
    super(props);
    this.state = {
      inputValue: this.props.query || "",
      focused: false
    };

    this.focused = null;

    this.inputOnChange = this.inputOnChange.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.onEnterPress = this.onEnterPress.bind(this);
    this.close = this.close.bind(this);
    this.selectMatchItem = this.selectMatchItem.bind(this);
  }

  close() {
    this.props.closeActiveSearch();
  }

  async onKeyDown(e: Element) {
    if (e.key !== "Enter") {
      return;
    }
    this.props.searchSources(this.state.inputValue);
  }

  onEnterPress() {
    if (this.focused) {
      const { setExpanded, file, expanded, match } = this.focused;
      if (setExpanded) {
        setExpanded(file, !expanded);
      } else {
        this.selectMatchItem(match);
      }
    }
  }

  componentWillUnmount() {
    const shortcuts = this.context.shortcuts;
    shortcuts.off("Enter", this.onEnterPress);
  }

  componentDidMount() {
    const shortcuts = this.context.shortcuts;
    shortcuts.on("Enter", this.onEnterPress);
  }

  inputOnChange(e) {
    const inputValue = e.target.value;
    this.setState({ inputValue });
  }

  selectMatchItem(matchItem) {
    this.props.selectSource(matchItem.sourceId, { line: matchItem.line });
  }

  getResults() {
    const { results } = this.props;
    return results
      .filter(result => result.filepath && result.matches.length > 0)
      .map(result => ({
        name: result.filepath,
        path: result.filepath,
        contents: result.matches
      }));
  }

  resultCount() {
    const results = this.getResults();
    return results.reduce(
      (count, file) => count + (file.contents ? file.contents.length : 0),
      0
    );
  }

  renderFile(
    file: Item,
    focused: boolean,
    expanded: boolean,
    setExpanded: () => void
  ) {
    if (focused) {
      this.focused = { setExpanded, file, expanded };
    }

    const svgArrow = Svg("arrow", {
      className: classnames({
        expanded
      })
    });

    const svgFile = Svg("file");

    const matches = ` (${file.contents.length} match${file.contents.length > 1
      ? "es"
      : ""})`;

    return (
      <div
        className={classnames("file-result", { focused })}
        key={file.sourceId}
        onClick={e => setExpanded(file, !expanded)}
      >
        {svgArrow}
        {svgFile}
        <span className="file-path">
          {getRelativePath(file.filepath)}
        </span>
        <span className="matches-summary">
          {matches}
        </span>
      </div>
    );
  }

  renderMatch(match: Item, focused: boolean) {
    if (focused) {
      this.focused = { match };
    }
    return (
      <div
        className={classnames("result", { focused })}
        onClick={() => setTimeout(() => this.selectMatchItem(match), 50)}
      >
        <span className="line-number" key={match.line}>
          {match.line}
        </span>
        {this.renderMatchValue(match.value)}
      </div>
    );
  }

  renderMatchValue(value: string) {
    const { inputValue } = this.state;
    let match;
    const len = inputValue.length;
    let matchIndexes = [];
    let matches = [];
    const re = new RegExp(escapeRegExp(inputValue), "g");
    while ((match = re.exec(value)) !== null) {
      matchIndexes.push(match.index);
    }

    matchIndexes.forEach((matchIndex, index) => {
      if (matchIndex > 0 && index === 0) {
        matches.push(
          <span className="line-match">
            {value.slice(0, matchIndex)}
          </span>
        );
      }
      if (matchIndex > matchIndexes[index - 1] + len) {
        matches.push(
          <span className="line-match">
            {value.slice(matchIndexes[index - 1] + len, matchIndex)}
          </span>
        );
      }
      matches.push(
        <span className="query-match" key={index}>
          {value.substr(matchIndex, len)}
        </span>
      );
      if (index === matchIndexes.length - 1) {
        matches.push(
          <span className="line-match">
            {value.slice(matchIndex + len, value.length)}
          </span>
        );
      }
    });

    return dom.span({ className: "line-value" }, ...matches);
  }

  renderResults() {
    const results = this.getResults();
    function getFilePath(item: Match) {
      return item.filepath
        ? `${item.sourceId}`
        : `${item.sourceId}-${item.line}-${item.column}`;
    }

    const renderItem = (item, depth, focused, _, expanded, { setExpanded }) => {
      return item.filepath
        ? this.renderFile(item, focused, expanded, setExpanded)
        : this.renderMatch(item, focused);
    };

    return (
      <ManagedTree
        getRoots={() => results}
        getChildren={file => file.contents}
        itemHeight={20}
        autoExpand={1}
        autoExpandDepth={1}
        focused={results[0]}
        getParent={item => null}
        getPath={getFilePath}
        renderItem={renderItem}
        autoExpandAll={false}
      />
    );
  }

  renderInput() {
    const resultCount = this.resultCount();
    const summaryMsg = L10N.getFormatStr(
      "sourceSearch.resultsSummary1",
      resultCount
    );

    return (
      <SearchInput
        query={this.state.inputValue}
        count={resultCount}
        placeholder={L10N.getStr("projectTextSearch.placeholder")}
        size="big"
        summaryMsg={summaryMsg}
        onChange={e => this.inputOnChange(e)}
        onFocus={() => this.setState({ focused: true })}
        onBlur={() => this.setState({ focused: false })}
        onKeyDown={e => this.onKeyDown(e)}
        handleClose={this.close}
        ref="searchInput"
      />
    );
  }

  render() {
    const { searchBottomBar } = this.props;
    return (
      <div className="project-text-search">
        <div className="header">
          {this.renderInput()}
          {searchBottomBar}
        </div>
        {this.renderResults()}
      </div>
    );
  }
}

TextSearch.contextTypes = {
  shortcuts: PropTypes.object
};

TextSearch.displayName = "TextSearch";
