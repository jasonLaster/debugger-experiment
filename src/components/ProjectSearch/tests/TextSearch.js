import React from "react";
import { shallow } from "enzyme";
import TextSearch from "../TextSearch.js";

function render(overrides = {}) {
  const defaultProps = {
    results: [],
    query: null,
    closeActiveSearch: jest.fn(),
    searchSources: jest.fn(),
    selectSource: jest.fn(),
    searchBottomBar: <div />
  };

  const props = Object.assign({}, defaultProps, overrides);

  const component = shallow(<TextSearch {...props} />);
  return { component, props };
}

describe("TextSearch", () => {
  it("basic", () => {
    const { component } = render();
    expect(component).toMatchSnapshot();
  });

  it("results + query", () => {
    const { component } = render({
      query: "foo",
      results: [
        {
          filepath: "bar.js",
          matches: [
            {
              sourceId: "foo",
              line: 12,
              column: 12,
              match: "foo",
              value: "bar foo foo",
              text: "bar foo foo"
            },
            {
              sourceId: "foo",
              line: 13,
              column: 23,
              match: "foo",
              value: "bar foo foo",
              text: "bar foo foo"
            }
          ]
        }
      ]
    });
    expect(component).toMatchSnapshot();
  });
});
