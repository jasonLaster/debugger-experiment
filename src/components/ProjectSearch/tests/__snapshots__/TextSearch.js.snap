// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextSearch basic 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="project-text-search"
>
    <div
        className="header"
    >
        <SearchInput
            count={0}
            handleClose={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Find in files…"
            query=""
            showErrorEmoji={true}
            size="big"
            summaryMsg="0 results"
        />
        <div />
    </div>
    <ManagedTree
        autoExpand={1}
        autoExpandAll={false}
        autoExpandDepth={1}
        focused={undefined}
        getChildren={[Function]}
        getParent={[Function]}
        getPath={[Function]}
        getRoots={[Function]}
        itemHeight={20}
        renderItem={[Function]}
    />
</div>,
  "nodes": Array [
    <div
      className="project-text-search"
>
      <div
            className="header"
      >
            <SearchInput
                  count={0}
                  handleClose={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Find in files…"
                  query=""
                  showErrorEmoji={true}
                  size="big"
                  summaryMsg="0 results"
            />
            <div />
      </div>
      <ManagedTree
            autoExpand={1}
            autoExpandAll={false}
            autoExpandDepth={1}
            focused={undefined}
            getChildren={[Function]}
            getParent={[Function]}
            getPath={[Function]}
            getRoots={[Function]}
            itemHeight={20}
            renderItem={[Function]}
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <TextSearch
        closeActiveSearch={[Function]}
        query={null}
        results={Array []}
        searchBottomBar={<div />}
        searchSources={[Function]}
        selectSource={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": TextSearch {
        "_reactInternalInstance": [Circular],
        "close": [Function],
        "context": Object {
          "shortcuts": undefined,
        },
        "focused": null,
        "inputOnChange": [Function],
        "onEnterPress": [Function],
        "onKeyDown": [Function],
        "props": Object {
          "closeActiveSearch": [Function],
          "query": null,
          "results": Array [],
          "searchBottomBar": <div />,
          "searchSources": [Function],
          "selectSource": [Function],
        },
        "refs": Object {},
        "selectMatchItem": [Function],
        "state": Object {
          "focused": false,
          "inputValue": "",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <div
          className="project-text-search"
>
          <div
                    className="header"
          >
                    <SearchInput
                              count={0}
                              handleClose={[Function]}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              placeholder="Find in files…"
                              query=""
                              showErrorEmoji={true}
                              size="big"
                              summaryMsg="0 results"
                    />
                    <div />
          </div>
          <ManagedTree
                    autoExpand={1}
                    autoExpandAll={false}
                    autoExpandDepth={1}
                    focused={undefined}
                    getChildren={[Function]}
                    getParent={[Function]}
                    getPath={[Function]}
                    getRoots={[Function]}
                    itemHeight={20}
                    renderItem={[Function]}
          />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="project-text-search"
>
          <div
                    className="header"
          >
                    <SearchInput
                              count={0}
                              handleClose={[Function]}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              placeholder="Find in files…"
                              query=""
                              showErrorEmoji={true}
                              size="big"
                              summaryMsg="0 results"
                    />
                    <div />
          </div>
          <ManagedTree
                    autoExpand={1}
                    autoExpandAll={false}
                    autoExpandDepth={1}
                    focused={undefined}
                    getChildren={[Function]}
                    getParent={[Function]}
                    getPath={[Function]}
                    getRoots={[Function]}
                    itemHeight={20}
                    renderItem={[Function]}
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <TextSearch
    closeActiveSearch={[Function]}
    query={null}
    results={Array []}
    searchBottomBar={<div />}
    searchSources={[Function]}
    selectSource={[Function]}
/>,
}
`;

exports[`TextSearch results + query 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="project-text-search"
>
    <div
        className="header"
    >
        <SearchInput
            count={2}
            handleClose={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Find in files…"
            query="foo"
            showErrorEmoji={true}
            size="big"
            summaryMsg="2 results"
        />
        <div />
    </div>
    <ManagedTree
        autoExpand={1}
        autoExpandAll={false}
        autoExpandDepth={1}
        focused={
            Object {
                "contents": Array [
                  Object {
                    "column": 12,
                    "line": 12,
                    "match": "foo",
                    "sourceId": "foo",
                    "text": "bar foo foo",
                    "value": "bar foo foo",
                  },
                  Object {
                    "column": 23,
                    "line": 13,
                    "match": "foo",
                    "sourceId": "foo",
                    "text": "bar foo foo",
                    "value": "bar foo foo",
                  },
                ],
                "name": "bar.js",
                "path": "bar.js",
              }
        }
        getChildren={[Function]}
        getParent={[Function]}
        getPath={[Function]}
        getRoots={[Function]}
        itemHeight={20}
        renderItem={[Function]}
    />
</div>,
  "nodes": Array [
    <div
      className="project-text-search"
>
      <div
            className="header"
      >
            <SearchInput
                  count={2}
                  handleClose={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Find in files…"
                  query="foo"
                  showErrorEmoji={true}
                  size="big"
                  summaryMsg="2 results"
            />
            <div />
      </div>
      <ManagedTree
            autoExpand={1}
            autoExpandAll={false}
            autoExpandDepth={1}
            focused={
                  Object {
                        "contents": Array [
                          Object {
                            "column": 12,
                            "line": 12,
                            "match": "foo",
                            "sourceId": "foo",
                            "text": "bar foo foo",
                            "value": "bar foo foo",
                          },
                          Object {
                            "column": 23,
                            "line": 13,
                            "match": "foo",
                            "sourceId": "foo",
                            "text": "bar foo foo",
                            "value": "bar foo foo",
                          },
                        ],
                        "name": "bar.js",
                        "path": "bar.js",
                      }
            }
            getChildren={[Function]}
            getParent={[Function]}
            getPath={[Function]}
            getRoots={[Function]}
            itemHeight={20}
            renderItem={[Function]}
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <TextSearch
        closeActiveSearch={[Function]}
        query="foo"
        results={
                Array [
                        Object {
                          "filepath": "bar.js",
                          "matches": Array [
                            Object {
                              "column": 12,
                              "line": 12,
                              "match": "foo",
                              "sourceId": "foo",
                              "text": "bar foo foo",
                              "value": "bar foo foo",
                            },
                            Object {
                              "column": 23,
                              "line": 13,
                              "match": "foo",
                              "sourceId": "foo",
                              "text": "bar foo foo",
                              "value": "bar foo foo",
                            },
                          ],
                        },
                      ]
        }
        searchBottomBar={<div />}
        searchSources={[Function]}
        selectSource={[Function]}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": TextSearch {
        "_reactInternalInstance": [Circular],
        "close": [Function],
        "context": Object {
          "shortcuts": undefined,
        },
        "focused": null,
        "inputOnChange": [Function],
        "onEnterPress": [Function],
        "onKeyDown": [Function],
        "props": Object {
          "closeActiveSearch": [Function],
          "query": "foo",
          "results": Array [
            Object {
              "filepath": "bar.js",
              "matches": Array [
                Object {
                  "column": 12,
                  "line": 12,
                  "match": "foo",
                  "sourceId": "foo",
                  "text": "bar foo foo",
                  "value": "bar foo foo",
                },
                Object {
                  "column": 23,
                  "line": 13,
                  "match": "foo",
                  "sourceId": "foo",
                  "text": "bar foo foo",
                  "value": "bar foo foo",
                },
              ],
            },
          ],
          "searchBottomBar": <div />,
          "searchSources": [Function],
          "selectSource": [Function],
        },
        "refs": Object {},
        "selectMatchItem": [Function],
        "state": Object {
          "focused": false,
          "inputValue": "foo",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <div
          className="project-text-search"
>
          <div
                    className="header"
          >
                    <SearchInput
                              count={2}
                              handleClose={[Function]}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              placeholder="Find in files…"
                              query="foo"
                              showErrorEmoji={true}
                              size="big"
                              summaryMsg="2 results"
                    />
                    <div />
          </div>
          <ManagedTree
                    autoExpand={1}
                    autoExpandAll={false}
                    autoExpandDepth={1}
                    focused={
                              Object {
                                        "contents": Array [
                                          Object {
                                            "column": 12,
                                            "line": 12,
                                            "match": "foo",
                                            "sourceId": "foo",
                                            "text": "bar foo foo",
                                            "value": "bar foo foo",
                                          },
                                          Object {
                                            "column": 23,
                                            "line": 13,
                                            "match": "foo",
                                            "sourceId": "foo",
                                            "text": "bar foo foo",
                                            "value": "bar foo foo",
                                          },
                                        ],
                                        "name": "bar.js",
                                        "path": "bar.js",
                                      }
                    }
                    getChildren={[Function]}
                    getParent={[Function]}
                    getPath={[Function]}
                    getRoots={[Function]}
                    itemHeight={20}
                    renderItem={[Function]}
          />
</div>,
        "_debugID": 4,
        "_renderedOutput": <div
          className="project-text-search"
>
          <div
                    className="header"
          >
                    <SearchInput
                              count={2}
                              handleClose={[Function]}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              placeholder="Find in files…"
                              query="foo"
                              showErrorEmoji={true}
                              size="big"
                              summaryMsg="2 results"
                    />
                    <div />
          </div>
          <ManagedTree
                    autoExpand={1}
                    autoExpandAll={false}
                    autoExpandDepth={1}
                    focused={
                              Object {
                                        "contents": Array [
                                          Object {
                                            "column": 12,
                                            "line": 12,
                                            "match": "foo",
                                            "sourceId": "foo",
                                            "text": "bar foo foo",
                                            "value": "bar foo foo",
                                          },
                                          Object {
                                            "column": 23,
                                            "line": 13,
                                            "match": "foo",
                                            "sourceId": "foo",
                                            "text": "bar foo foo",
                                            "value": "bar foo foo",
                                          },
                                        ],
                                        "name": "bar.js",
                                        "path": "bar.js",
                                      }
                    }
                    getChildren={[Function]}
                    getParent={[Function]}
                    getPath={[Function]}
                    getRoots={[Function]}
                    itemHeight={20}
                    renderItem={[Function]}
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <TextSearch
    closeActiveSearch={[Function]}
    query="foo"
    results={
        Array [
            Object {
              "filepath": "bar.js",
              "matches": Array [
                Object {
                  "column": 12,
                  "line": 12,
                  "match": "foo",
                  "sourceId": "foo",
                  "text": "bar foo foo",
                  "value": "bar foo foo",
                },
                Object {
                  "column": 23,
                  "line": 13,
                  "match": "foo",
                  "sourceId": "foo",
                  "text": "bar foo foo",
                  "value": "bar foo foo",
                },
              ],
            },
          ]
    }
    searchBottomBar={<div />}
    searchSources={[Function]}
    selectSource={[Function]}
/>,
}
`;
